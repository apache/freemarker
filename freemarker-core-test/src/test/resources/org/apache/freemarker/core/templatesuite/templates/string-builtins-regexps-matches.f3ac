<#--
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->
<#assign input>
L16
L27
L38
L49
</#assign>

List mode:
<#assign matches = input?matches(".+") >
Size: ${matches?size}
<#list 0..<matches?size as i>[${matches[i]}]</#list>
<#list matches as m>[${m}]</#list>

Iterator mode:
<#assign matches = input?matches(".+") >
<#list matches as m>[${m}]</#list>
<#list matches as m>[${m}(${matches?join(', ')})]</#list>
<#list matches as m>[${m}]</#list>

Iterator mode changes to list mode:
<#assign matches = input?matches(".+") >
<#list matches as m>[${m}]/${matches?size}</#list>
<#list matches as m>[${m}]</#list>

Iterator mode changes to list mode 2:
<#assign matches = input?matches(".+") >
<#list matches as m>[${m}]</#list>
<#list matches as m>[${m}]/${matches?size}<#t></#list>

List mode with embedded iteration:
<#assign matches = input?matches(".+") >
<#list 0..<matches?size as i>[${matches[i]}(${matches?join(', ')})]</#list>

Entire input match:
<#assign matches = input?matches(r".*(\d)(\d)") >
<#assign firstGS = false>
<#list matches as m>
- M: ${m}
    <#if firstGS?isBoolean>
      <#assign firstGS = m?groups>
    </#if>
    <#list m?groups as g>
    - G: ${g}
    </#list>
</#list>
firstGS was: ${firstGS?join(', ')}

Entire input match 2:
<#assign match = "x12"?matches(r"x(\d)(\d)") >
Matches: ${match?c}
<#list match?groups as g>
- G: ${g}
</#list>
As list:
<#list match as m>
- M: ${m}
  <#list m?groups as g>
    - G: ${g}
  </#list>
</#list>
Groups again:
<#list match?groups as g>
- G: ${g}
</#list>

Entire input match 3:
<#assign match = "x12"?matches(r"y(\d)(\d)") >
Matches: ${match?c}
<@assertEquals expected=3 actual=match?groups?size />
<@assertEquals expected=0 actual=match?size />

Entire input match 4:
<#assign match = "x12"?matches(r"x(\d)(\d)") >
Matches: ${match?c}
<#assign gs = match?groups>
<@assertEquals expected=3 actual=gs?size />
<@assertEquals expected=1 actual=match?size />
- G: ${gs[0]}
- G: ${match?groups[1]}
- G: ${gs[2]}

Substring match nested into entire input match:
<#assign match = "x12"?matches(r"x(\d)(\d)") >
<#list match?groups as g>
- G: ${g} (<#list match as m>[${m}{${m?groups?join(', ')}}]</#list>)
</#list>

Different entire input and substring matches:
<#assign match = "123"?matches(r"(\d+?)") >
${match?groups?join(", ")}
<#list match as m>
- M: ${m} (Gs: ${m?groups?join(", ")})
</#list>

Different entire input and substring matches 2:
<#assign match = "123"?matches(r"\d+?") >
${match?groups?join(", ")}
<#list match as m>
- M: ${m} (Gs: ${m?groups?join(", ")})
</#list>