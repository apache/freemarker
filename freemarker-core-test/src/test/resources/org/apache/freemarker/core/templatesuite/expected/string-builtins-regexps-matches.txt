/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

List mode:
Size: 4
[L16][L27][L38][L49]
[L16][L27][L38][L49]

Iterator mode:
[L16][L27][L38][L49]
[L16(L16, L27, L38, L49)][L27(L16, L27, L38, L49)][L38(L16, L27, L38, L49)][L49(L16, L27, L38, L49)]
[L16][L27][L38][L49]

Iterator mode changes to list mode:
[L16]/4[L27]/4[L38]/4[L49]/4
[L16][L27][L38][L49]

Iterator mode changes to list mode 2:
[L16][L27][L38][L49]
[L16]/4[L27]/4[L38]/4[L49]/4

List mode with embedded iteration:
[L16(L16, L27, L38, L49)][L27(L16, L27, L38, L49)][L38(L16, L27, L38, L49)][L49(L16, L27, L38, L49)]

Entire input match:
- M: L16
    - G: L16
    - G: 1
    - G: 6
- M: L27
    - G: L27
    - G: 2
    - G: 7
- M: L38
    - G: L38
    - G: 3
    - G: 8
- M: L49
    - G: L49
    - G: 4
    - G: 9
firstGS was: L16, 1, 6

Entire input match 2:
Matches: true
- G: x12
- G: 1
- G: 2
As list:
- M: x12
    - G: x12
    - G: 1
    - G: 2
Groups again:
- G: x12
- G: 1
- G: 2

Entire input match 3:
Matches: false

Entire input match 4:
Matches: true
- G: x12
- G: 1
- G: 2

Substring match nested into entire input match:
- G: x12 ([x12{x12, 1, 2}])
- G: 1 ([x12{x12, 1, 2}])
- G: 2 ([x12{x12, 1, 2}])

Different entire input and substring matches:
123, 123
- M: 1 (Gs: 1, 1)
- M: 2 (Gs: 2, 2)
- M: 3 (Gs: 3, 3)

Different entire input and substring matches 2:
123
- M: 1 (Gs: 1)
- M: 2 (Gs: 2)
- M: 3 (Gs: 3)
